File: server/keychain.js
'use strict';

/**
 * Cross-platform secret storage backed by the OS keychain.
 *
 * Refresh tokens are long-lived credentials and must not sit in a plaintext
 * file. Each platform gets its native store; if none is reachable we throw
 * rather than silently degrading to disk. Node standard library only.
 *
 * Caveat on macOS: the `security` CLI takes the secret as an argument, so it
 * is briefly visible in `ps` output on a multi-user machine. The Linux and
 * Windows backends both pass secrets over stdin and do not have this exposure.
 */

const { spawn } = require('child_process');
const fs = require('fs');
const { cleanEnv } = require('./env');

/**
 * Resolve a helper binary to an absolute path when we know where it lives.
 * Claude Desktop's Code/Cowork session hosts run this bridge with a stripped
 * environment in which PATH is missing, so a bare `spawn('security')` fails
 * with ENOENT there — every keychain read then looks like "no credentials"
 * and every start triggers a fresh sign-in.
 */
function resolveBinary(name) {
  const known = {
    security: ['/usr/bin/security'],
    osascript: ['/usr/bin/osascript'],
    open: ['/usr/bin/open'],
    'secret-tool': ['/usr/bin/secret-tool', '/usr/local/bin/secret-tool'],
    powershell: [
      `${process.env.SystemRoot || 'C:\\Windows'}\\System32\\WindowsPowerShell\\v1.0\\powershell.exe`,
    ],
  };
  for (const candidate of known[name] || []) {
    try {
      fs.accessSync(candidate, fs.constants.X_OK);
      return candidate;
    } catch (_) { /* try the next */ }
  }
  return name; // fall back to PATH lookup
}

// Overridable so tests can use a throwaway service instead of the real one.
const SERVICE = cleanEnv('GIGASHEET_KEYCHAIN_SERVICE', 'com.gigasheet.claude-extension');

class KeychainError extends Error {}

function run(cmd, args, { stdin = null, check = true, timeoutMs = 15000 } = {}) {
  return new Promise((resolve, reject) => {
    let settled = false;
    const finish = (fn, value) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      fn(value);
    };

    let child;
    try {
      child = spawn(resolveBinary(cmd), args, { stdio: ['pipe', 'pipe', 'pipe'], windowsHide: true });
    } catch (err) {
      reject(new KeychainError(`${cmd} could not be started: ${err.message}`));
      return;
    }

    const timer = setTimeout(() => {
      child.kill();
      finish(reject, new KeychainError(`${cmd} timed out`));
    }, timeoutMs);

    let stdout = '';
    let stderr = '';
    child.stdout.on('data', (d) => { stdout += d; });
    child.stderr.on('data', (d) => { stderr += d; });
    child.on('error', (err) => {
      finish(reject, new KeychainError(
        err.code === 'ENOENT' ? `${cmd} is not installed` : `${cmd} failed: ${err.message}`
      ));
    });
    child.on('close', (code) => {
      if (check && code !== 0) {
        finish(reject, new KeychainError(
          `${cmd} failed (exit ${code}): ${stderr.trim().slice(0, 200)}`
        ));
      } else {
        finish(resolve, { code, stdout, stderr });
      }
    });

    child.stdin.on('error', () => {});
    child.stdin.end(stdin === null ? undefined : stdin);
  });
}

// --------------------------------------------------------------------------
// macOS
// --------------------------------------------------------------------------

const darwin = {
  async set(account, secret) {
    await run('security', [
      'add-generic-password',
      '-U', // update if it already exists
      '-a', account,
      '-s', SERVICE,
      '-w', secret,
    ]);
  },
  async get(account) {
    const r = await run('security', [
      'find-generic-password', '-a', account, '-s', SERVICE, '-w',
    ], { check: false });
    if (r.code !== 0) return null; // not found is the common case, not an error
    return r.stdout.trim() || null;
  },
  async del(account) {
    await run('security', [
      'delete-generic-password', '-a', account, '-s', SERVICE,
    ], { check: false });
  },
};

// --------------------------------------------------------------------------
// Linux (libsecret)
// --------------------------------------------------------------------------

const linux = {
  async set(account, secret) {
    await run('secret-tool', [
      'store', '--label=Gigasheet Claude extension',
      'service', SERVICE, 'account', account,
    ], { stdin: secret });
  },
  async get(account) {
    const r = await run('secret-tool', [
      'lookup', 'service', SERVICE, 'account', account,
    ], { check: false });
    if (r.code !== 0) return null;
    return r.stdout.trim() || null;
  },
  async del(account) {
    await run('secret-tool', [
      'clear', 'service', SERVICE, 'account', account,
    ], { check: false });
  },
};

// --------------------------------------------------------------------------
// Windows (DPAPI via PowerShell, scoped to the current user)
// --------------------------------------------------------------------------

const PS_STORE = (account) => `
$ErrorActionPreference = 'Stop'
Add-Type -AssemblyName System.Security
$secret = [Console]::In.ReadToEnd()
$bytes = [Text.Encoding]::UTF8.GetBytes($secret)
$enc = [Security.Cryptography.ProtectedData]::Protect($bytes, $null, 'CurrentUser')
$dir = Join-Path $env:LOCALAPPDATA 'Gigasheet'
New-Item -ItemType Directory -Force -Path $dir | Out-Null
[IO.File]::WriteAllBytes((Join-Path $dir '${account}.bin'), $enc)
`;

const PS_LOAD = (account) => `
$ErrorActionPreference = 'Stop'
Add-Type -AssemblyName System.Security
$path = Join-Path $env:LOCALAPPDATA 'Gigasheet\\${account}.bin'
if (-not (Test-Path $path)) { exit 1 }
$enc = [IO.File]::ReadAllBytes($path)
$bytes = [Security.Cryptography.ProtectedData]::Unprotect($enc, $null, 'CurrentUser')
[Console]::Out.Write([Text.Encoding]::UTF8.GetString($bytes))
`;

const PS_DELETE = (account) => `
$path = Join-Path $env:LOCALAPPDATA 'Gigasheet\\${account}.bin'
if (Test-Path $path) { Remove-Item $path -Force }
`;

function ps(script, opts = {}) {
  return run('powershell', ['-NoProfile', '-NonInteractive', '-Command', script], opts);
}

const win32 = {
  async set(account, secret) { await ps(PS_STORE(account), { stdin: secret }); },
  async get(account) {
    const r = await ps(PS_LOAD(account), { check: false });
    if (r.code !== 0) return null;
    return r.stdout.trim() || null;
  },
  async del(account) { await ps(PS_DELETE(account), { check: false }); },
};

// --------------------------------------------------------------------------
// Dispatch
// --------------------------------------------------------------------------

function backend() {
  switch (process.platform) {
    case 'darwin': return darwin;
    case 'linux': return linux;
    case 'win32': return win32;
    default: throw new KeychainError(`no keychain backend for platform ${process.platform}`);
  }
}

// Account names are fixed strings we control, never user input; they end up
// inside shell-free spawn args (and, on Windows, a file name), so keep them
// to a conservative character set anyway.
function checkAccount(account) {
  if (!/^[A-Za-z0-9_.-]{1,64}$/.test(account)) {
    throw new KeychainError(`invalid account name ${JSON.stringify(account)}`);
  }
}

async function setSecret(account, secret) {
  checkAccount(account);
  await backend().set(account, secret);
}

/** Resolves to the stored secret, or null if absent. */
async function getSecret(account) {
  checkAccount(account);
  return backend().get(account);
}

async function deleteSecret(account) {
  checkAccount(account);
  await backend().del(account);
}

/**
 * Check that this platform's keychain is usable right now. Resolves to null
 * when it is, or a short reason string when it is not.
 */
async function unavailableReason() {
  try {
    backend();
  } catch (err) {
    return err.message;
  }
  // Per-process name: Desktop launches two instances a second apart, and a
  // shared probe item let one delete the other's mid-check.
  const probe = `availability_probe_${process.pid}`;
  try {
    await setSecret(probe, 'ok');
    const ok = (await getSecret(probe)) === 'ok';
    await deleteSecret(probe);
    return ok ? null : 'probe value did not round-trip';
  } catch (err) {
    return err.message;
  }
}

/** True if this platform's keychain is usable right now. */
async function available() {
  return (await unavailableReason()) === null;
}

module.exports = {
  setSecret, getSecret, deleteSecret, available, unavailableReason, resolveBinary, KeychainError, SERVICE,
};

File: python-reference/keychain.py
"""Cross-platform secret storage backed by the OS keychain.

Refresh tokens are long-lived credentials and must not sit in a plaintext file.
Each platform gets its native store; if none is reachable we raise rather than
silently degrading to disk. Standard library only.

Caveat on macOS: the `security` CLI takes the secret as an argument, so it is
briefly visible in `ps` output on a multi-user machine. The Linux and Windows
backends both pass secrets over stdin and do not have this exposure. Replacing
the macOS path with a ctypes call into Security.framework would close it.
"""

import subprocess
import sys

SERVICE = "com.gigasheet.claude-extension"


class KeychainError(Exception):
    """The platform keychain could not be reached or refused the operation."""


def _run(cmd, stdin_data=None, check=True):
    try:
        proc = subprocess.run(
            cmd,
            input=stdin_data,
            capture_output=True,
            text=True,
            timeout=15,
        )
    except FileNotFoundError:
        raise KeychainError("%s is not installed" % cmd[0])
    except subprocess.TimeoutExpired:
        raise KeychainError("%s timed out" % cmd[0])

    if check and proc.returncode != 0:
        raise KeychainError(
            "%s failed (exit %d): %s"
            % (cmd[0], proc.returncode, proc.stderr.strip()[:200])
        )
    return proc


# --------------------------------------------------------------------------
# macOS
# --------------------------------------------------------------------------

def _darwin_set(account, secret):
    _run([
        "security", "add-generic-password",
        "-U",                # update if it already exists
        "-a", account,
        "-s", SERVICE,
        "-w", secret,
    ])


def _darwin_get(account):
    proc = _run([
        "security", "find-generic-password",
        "-a", account, "-s", SERVICE, "-w",
    ], check=False)
    if proc.returncode != 0:
        return None  # not found is the common case, not an error
    return proc.stdout.strip() or None


def _darwin_delete(account):
    _run([
        "security", "delete-generic-password",
        "-a", account, "-s", SERVICE,
    ], check=False)


# --------------------------------------------------------------------------
# Linux (libsecret)
# --------------------------------------------------------------------------

def _linux_set(account, secret):
    _run([
        "secret-tool", "store",
        "--label=Gigasheet Claude extension",
        "service", SERVICE, "account", account,
    ], stdin_data=secret)


def _linux_get(account):
    proc = _run([
        "secret-tool", "lookup", "service", SERVICE, "account", account,
    ], check=False)
    if proc.returncode != 0:
        return None
    return proc.stdout.strip() or None


def _linux_delete(account):
    _run([
        "secret-tool", "clear", "service", SERVICE, "account", account,
    ], check=False)


# --------------------------------------------------------------------------
# Windows (DPAPI via PowerShell, scoped to the current user)
# --------------------------------------------------------------------------

_PS_STORE = r"""
$ErrorActionPreference = 'Stop'
Add-Type -AssemblyName System.Security
$secret = [Console]::In.ReadToEnd()
$bytes = [Text.Encoding]::UTF8.GetBytes($secret)
$enc = [Security.Cryptography.ProtectedData]::Protect($bytes, $null, 'CurrentUser')
$dir = Join-Path $env:LOCALAPPDATA 'Gigasheet'
New-Item -ItemType Directory -Force -Path $dir | Out-Null
[IO.File]::WriteAllBytes((Join-Path $dir '%s.bin'), $enc)
"""

_PS_LOAD = r"""
$ErrorActionPreference = 'Stop'
Add-Type -AssemblyName System.Security
$path = Join-Path $env:LOCALAPPDATA 'Gigasheet\%s.bin'
if (-not (Test-Path $path)) { exit 1 }
$enc = [IO.File]::ReadAllBytes($path)
$bytes = [Security.Cryptography.ProtectedData]::Unprotect($enc, $null, 'CurrentUser')
[Console]::Out.Write([Text.Encoding]::UTF8.GetString($bytes))
"""

_PS_DELETE = r"""
$path = Join-Path $env:LOCALAPPDATA 'Gigasheet\%s.bin'
if (Test-Path $path) { Remove-Item $path -Force }
"""


def _ps(script, stdin_data=None, check=True):
    return _run(
        ["powershell", "-NoProfile", "-NonInteractive", "-Command", script],
        stdin_data=stdin_data,
        check=check,
    )


def _windows_set(account, secret):
    _ps(_PS_STORE % account, stdin_data=secret)


def _windows_get(account):
    proc = _ps(_PS_LOAD % account, check=False)
    if proc.returncode != 0:
        return None
    return proc.stdout.strip() or None


def _windows_delete(account):
    _ps(_PS_DELETE % account, check=False)


# --------------------------------------------------------------------------
# Dispatch
# --------------------------------------------------------------------------

_BACKENDS = {
    "darwin": (_darwin_set, _darwin_get, _darwin_delete),
    "linux": (_linux_set, _linux_get, _linux_delete),
    "win32": (_windows_set, _windows_get, _windows_delete),
}


def _backend():
    for prefix, funcs in _BACKENDS.items():
        if sys.platform.startswith(prefix):
            return funcs
    raise KeychainError("no keychain backend for platform %r" % sys.platform)


def set_secret(account, secret):
    _backend()[0](account, secret)


def get_secret(account):
    """Returns the stored secret, or None if absent."""
    return _backend()[1](account)


def delete_secret(account):
    _backend()[2](account)


def available():
    """True if this platform's keychain is usable right now."""
    try:
        _backend()
    except KeychainError:
        return False
    probe = "__availability_probe__"
    try:
        set_secret(probe, "ok")
        result = get_secret(probe) == "ok"
        delete_secret(probe)
        return result
    except KeychainError:
        return False
